src/modules/roc_netio/target_uv/roc_netio/udp_receiver.cpp

```diff
@@ -42,6 +42,13 @@ void UDPReceiver::destroy() {
 }
 
 bool UDPReceiver::start(packet::Address& bind_address) {
+    if (bind_address.port() > 0 && bind_address.port() < 1024) {
+        roc_log(LogError,
+                "udp receiver: can't open port %s: ports (0, 1024] are reserved",
```

Comment:
Umm.. This is wrong. Ports below 1024 are not "reserved", they are privileged. You *can* bind to such a port, but this requires root or CAP_NET_BIND_SERVICE. This is up to the user to choose ports and privileges and we should not force any policy here. Instead, we should correctly handle errors from libuv and exit with non-zero code instead of a panic.

Reply:
Did you try to run ROC under the root user? I tried and it does not work. As a result I considered that we shouldn't allow user to choose a port number in this range.

Reply:
If it doesn't work, it's a bug. You can test udp ports below 1024 with sudo nc, for example.

Reply:
Just checked - works for me, e.g.:

```
sudo roc-recv -vv -s rtp+rs8m::101 -r rs8m::102 -d wav -o /dev/null
```

What error did you get? Probably you were running roc-recv under sudo and pulseaudio server was not running under the root user?

Reply:
>  Probably you were running roc-recv under sudo and pulseaudio server was not running under the root user

Correct

Reply:
> Instead, we should correctly handle errors from libuv

Reworked.

We need to ensure that uv HANDLE will be properly closed on any errors
during UDPReceiver::start call.

We did not release it if there is an error on:

* uv_udp_bind
* uv_udp_getsockname

Reply:
### Problem 1

Unfortunately, it seems that you didn't fix the problem.

First, take a look at this code:

```
bool Transceiver::add_udp_receiver_(Task& task) {
    ...

    core::SharedPtr<UDPReceiver> rp = new (allocator_)
        UDPReceiver(loop_, *task.writer, packet_pool_, buffer_pool_, allocator_);

    ...

    if (!rp->start(*task.address)) {
        roc_log(LogError, "transceiver: can't add port %s: can't start receiver",
                packet::address_to_str(*task.address).c_str());
        return false;
    }

    receivers_.push_back(*rp);
    num_ports_++;

    return true;
}
```

If UDPReceiver::start() returns false, the UDPReceiver object is immediately destroyed.

Second, close operations are asynchronous in libuv. When you call UDPReceiver::close_() and it calls uv_close(), the handle is not closed immediately. Instead, it is marked as *closing* and will be fully closed in the next event loop iteration.

This means that in case of an error we destroy an object that is still used by libuv.

### Problem 2

Don't we have the same problem in UDPSender?

### Problem 3

I've looked at the trx and receiver code in more detail and found another related problem (not caused by this PR):

```
bool Transceiver::remove_port_(Task& task) {
    for (core::SharedPtr<UDPReceiver> rp = receivers_.front(); rp;
         rp = receivers_.nextof(*rp)) {
        if (rp->address() == *task.address) {
            rp->remove(receivers_);
            num_ports_--;
            return true;
        }
    }

    ...
}

void UDPReceiver::remove(core::List<UDPReceiver>& container) {
    roc_panic_if(container_);

    if (handle_initialized_) {
        stop();
        container_ = &container;
        address_ = packet::Address();
    } else {
        container.remove(*this);
    }
}
```

This piece correctly handles asynchronous closing, but it has a problem. It schedules the receiver for removing, keeps it in the list, and then immediately returns success to the caller.

Then the receiver will be closed in the background and removed from the list shortly. However, if the caller will be fast enough to try to rebind this port before the old receiver is closed, it will get an unexpected error because such a port is still in the list.

So there is a race. We report that the port is removed while it is not yet removed. Instead, we should wait until it is removed before returning success to the user.

### Solution

I suggest to address all these problems simultaneously.

I suggest to introduce a new list for pending-removal objects in trx. This how it should be used:

* when an object is created and started successfully, it is added to receivers_ or senders_ list, as before
* when an object is created but can't be started, asynchronous close is issued and the object is added to pending_removal_
* when a port is removed by user, asynchronous close is issued and the object is moved to pending_removal_
* UDPReceiver::remove and UDPSender::remove should accept a callback argument instead of a container; they should call that callback when the asynchronous close is finished
* Transceiver should pass them a callback that removes the object from pending_removal_ and signals the condition variable
* Transceiver::remove_port() should first wait until the task is finished and then wait until the asynchronous close is finished by looping on the condition an checking the pending_removal_ list

Reply:
Also, we need unit tests for the case when bind fails. The simplest way to reproduce this error is to specify an already bound port, which can be achieved by creating the second trx and binding the same port on it.

BTW, if we had such tests already, they would detect the problem in this PR when running under clang sanitizers.

Reply:
Oh..It took quite a long time.

----

* Transceiver::start(), Transceiver::join() were removed,
  now Transceiver::stop() waits all senders and receivers
  to be removed and joins the background thread.

* Transceiver automatically calls Transceiver::stop() in
  destructor if it was not called explicitly.

* Transceiver::remove_port waits until port will be actually removed.

* Transceiver::add_udp_receiver, Transceiver::add_udp_sender 
  Transceiver::remove_port should not be called after Transceiver::stop(). 
  It simplifies the internals but makes the API more strict.

* UDPSender::remove, UPDReceiver::remove were removed.